storm-core/src/jvm/org/apache/storm/StormTimer.java

```diff
@@ -0,0 +1,274 @@
+                QueueEntry queueEntry = null;
+                try {
+                    synchronized (this.lock) {
+                        queueEntry = this.queue.peek();
+                    }
+                    if ((queueEntry != null) && (Time.currentTimeMillis() >= queueEntry.endTimeMs)) {
+                        // It is imperative to not run the function
+                        // inside the timer lock. Otherwise, it is
+                        // possible to deadlock if the fn deals with
+                        // other locks, like the submit lock.
+                        synchronized (this.lock) {
+                            this.queue.poll();
+                        }
+                        queueEntry.afn.run(null);
+                    } else if (queueEntry != null) {
+                        //  If any events are scheduled, sleep until
+                        // event generation. If any recurring events
+                        // are scheduled then we will always go
+                        // through this branch, sleeping only the
+                        // exact necessary amount of time. We give
+                        // an upper bound, e.g. 1000 millis, to the
+                        // sleeping time, to limit the response time
+                        // for detecting any new event within 1 secs.
+                        Time.sleep(Math.min(1000, (queueEntry.endTimeMs - Time.currentTimeMillis())));
+                    } else {
+                        // Otherwise poll to see if any new event
+                        // was scheduled. This is, in essence, the
+                        // response time for detecting any new event
+                        // schedulings when there are no scheduled
+                        // events.
+                        Time.sleep(1000);
+                    }
+                } catch (Throwable t) {
+                    if (!(Utils.exceptionCauseIsInstanceOf(InterruptedException.class, t))) {
+                        this.onKill.run(t);
+                        this.setActive(false);
+                        throw new RuntimeException(t);
+                    }
+                }
+            }
+            this.cancelNotifier.release();
+        }
+
+        public void setOnKillFunc(TimerFunc onKill) {
+            this.onKill = onKill;
+        }
+
+        public void setActive(boolean flag) {
+            this.active.set(flag);
+        }
+
+        public boolean isActive() {
+            return this.active.get();
+        }
+
+        public void add(QueueEntry queueEntry) {
+            this.queue.add(queueEntry);
+        }
+    }
+
+    /**
+     * Makes a Timer in the form of a StormTimerTask Object
+     * @param name name of the timer
+     * @param onKill function to call when timer is killed unexpectedly
+     * @return StormTimerTask object that was initialized
+     */
+    public static StormTimerTask mkTimer(String name, TimerFunc onKill) {
```

Comment:
Why not make this a Constructor, and hide StormTimerTask as an implementation detail?